### ISSUE TITLE
Sandbox not used with distributed artifact caching

### ISSUE THREAD

**Issue by dfabulich** [bazelbuild/bazel#1531]
https://groups.google.com/forum/#!topic/bazel-discuss/GINp6C6tZ4I

**Comment by ulfjack** [comment]
@hhclam

**Comment by philwo** [comment]
I want to solve this some day in the future with something like a "Spawn scheduler" where individual strategies can just say "Sorry, I can't handle this, please execute the spawn with the next fallback strategy". This would get rid of all those hardcoded references to StandaloneSpawnStrategy from others.

In the meantime, it could probably easily be fixed by checking whether sandboxing is supported in the remote execution strategy and then using the appropriate sandboxing strategy instead of StandaloneSpawnStrategy.

**Comment by ulfjack** [comment]
Moving this up to P1 and targeting 0.6.0; I think this is the most important missing feature for remote caching. I already have some of the necessary changes in a local branch, but it's not quite there yet.

**Comment by philwo** [comment]
@ulfjack I agree with the increased priority. Let's get this done. :)

**Comment by hhclam** [comment]
We have started using this feature to speed up builds. What's the latest update? I would love to see this works with sandbox and java worker process.

**Comment by ulfjack** [comment]
Before I left for two weeks of vacation, I merged 9274cba2540d1d1c7824147f1d999ac785eeed85, which adds --experimental_remote_spawn_cache, which enables support for remote caching with any local execution strategy, including sandboxing and workers. I still need to update the documentation, but you basically set --remote_cache and --experimental_remote_spawn_cache and that should work. It isn't currently possible to disable remote caching, but keep local caching, and you may also need to set --experimental_strict_action_env to get cache hits across users / machines.

**Comment by ulfjack** [comment]
Also see #2574 for the cross-user caching issue.

**Comment by mmorearty** [comment]
With `--experimental_remote_spawn_cache` I see Bazel attempting to _download_ artifacts from the cache, but I never see it attempting to _upload_.

I have a simple BUILD file with a genrule named `"main"`. My command line (with Bazel 0.8.1):

```bash
bazel build \
  --experimental_remote_spawn_cache \
  --remote_rest_cache=http://localhost:7777 \
  --spawn_strategy=sandboxed \
  -s \
  :main
```

When I look at the logs for the REST cache that I have running at `http://localhost:7777`, I see `GET` requests but no `PUT` requests.

**Comment by hhclam** [comment]
Strange.. Make sure you have this flag set to true. https://github.com/bazelbuild/bazel/blob/ba66e72295d763e43b92179ce6bd46476084a548/src/main/java/com/google/devtools/build/lib/remote/RemoteOptions.java#L102
It should be true by default though. Also sandbox is on by default and you don't need to specify it.

**Comment by mmorearty** [comment]
Sorry, my mistake... not sure what I did wrong. Uploading is working fine. (And thanks for pointing out that sandboxing is on by default)

**Comment by buchgr** [comment]
Can this be closed?

**Comment by ulfjack** [comment]
I want to enable the new behavior by default, and for that it needs to actually play nicely with remote execution. We could track that in a separate issue, but I think it's better to use this one.

**Comment by ulfjack** [comment]
Fixed in cdf118a607ab179807cd39d0cafdb551859d1cd1.
